source/iterator/snapshot.go

```diff
@@ -0,0 +1,189 @@
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jmoiron/sqlx"
+)
+
+// Snapshot represents an implementation of a Snapshot iterator for Oracle.
+type Snapshot struct {
+	repo     *repository.Oracle
+	position *Position
+
+	// table represents a table name
+	table string
+	// snapshotTable represents a name of snapshot table
+	snapshotTable string
+	// keyColumn represents a name of column what iterator use for setting key in record
+	keyColumn string
+	// orderingColumn represents a name of column what iterator use for sorting data
+	orderingColumn string
+	// columns represents a list of table's columns for record payload.
+	// if empty - will get all columns
+	columns []string
+	// batchSize represents a size of batch
+	batchSize int
+
+	rows *sqlx.Rows
+	// columnTypes represents a columns' description from table
+	columnTypes map[string]coltypes.ColumnDescription
+}
+
+// SnapshotParams represents an incoming params for the NewSnapshot function.
+type SnapshotParams struct {
+	Repo           *repository.Oracle
+	Position       *Position
+	Table          string
+	SnapshotTable  string
+	KeyColumn      string
+	OrderingColumn string
+	Columns        []string
+	BatchSize      int
+	ColumnTypes    map[string]coltypes.ColumnDescription
+}
+
+// NewSnapshot creates a new instance of the Snapshot iterator.
+func NewSnapshot(ctx context.Context, params SnapshotParams) (*Snapshot, error) {
+	iterator := &Snapshot{
+		repo:           params.Repo,
+		position:       params.Position,
+		table:          params.Table,
+		snapshotTable:  params.SnapshotTable,
+		keyColumn:      params.KeyColumn,
+		orderingColumn: params.OrderingColumn,
+		columns:        params.Columns,
+		batchSize:      params.BatchSize,
+		columnTypes:    params.ColumnTypes,
+	}
+
+	err := iterator.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	return iterator, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (i *Snapshot) HasNext(ctx context.Context) (bool, error) {
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	if err := i.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	return i.rows.Next(), nil
+}
+
+// Next returns the next record.
+func (i *Snapshot) Next(_ context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := i.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := coltypes.TransformRow(row, i.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[i.orderingColumn]; !ok {
+		return sdk.Record{}, errOrderingColumnIsNotExist
+	}
+
+	i.position = &Position{
+		Mode: ModeSnapshot,
+		// set the value from i.orderingColumn column you chose
+		LastProcessedVal: transformedRow[i.orderingColumn],
+	}
```

Comment:
I'm wondering, if we would have problems, if we set the position, but we fail below, when marshalling the position or transforming the row.

We probably shouldn't set the position, if we cannot marshal it. Also, if the row cannot be transformed, then we cannot return it, so `i.position` won't really be the current position.

Reply:
Yep, it would be better.

I've moved the 

So I set a new position into the variable, to avoid saving the position into the struct until we successfully marshal the position.